documentation/modules/oauth/proc-oauth-authorization-keycloak-example.adoc

```diff
@@ -0,0 +1,641 @@
+//
+// assembly-oauth-authorization.adoc
+
+[id='proc-oauth-authorization-keycloak-example_{context}']
+= Trying Keycloak Authorization Services
+
+[role="_abstract"]
+This example explains how to use Keycloak Authorization Services with `keycloak` authorization.
+Use Keycloak Authorization Services to enforce access restrictions on Kafka clients.
+Keycloak Authorization Services use authorization scopes, policies and permissions to define and apply access control to resources.
+
+Keycloak Authorization Services REST endpoints provide a list of granted permissions on resources for authenticated users.
+The list of grants (permissions) is fetched from the Keycloak server as the first action after an authenticated session is established by the Kafka client.
+The list is refreshed in the background so that changes to the grants are detected.
+Grants are cached and enforced locally on the Kafka broker for each user session to provide fast authorization decisions.
+
+Strimzi provides two example files with the deployment artifacts for setting up Keycloak:
+
+`kafka-ephemeral-oauth-single-keycloak-authz.yaml`:: An example `Kafka` custom resource configured for OAuth 2.0 token-based authorization using Keycloak.
+You can use the custom resource to deploy a Kafka cluster that uses `keycloak` authorization and token-based `oauth` authentication.
+
+`kafka-authz-realm.json`:: An example Keycloak realm configured with sample groups, users, roles and clients.
+You can import the realm into a Keycloak instance to set up fine-grained permissions to access Kafka.
+
+If you want to try the example with Keycloak, use these files to perform the tasks outlined in this section in the order shown.
+
+. xref:proc-oauth-authorization-keycloak-example-setup_{context}[Accessing the Keycloak Admin Console]
+. xref:proc-oauth-authorization-keycloak-example-deploy-kafka_{context}[Deploying a Kafka cluster with Keycloak authorization]
+. xref:proc-oauth-authorization-keycloak-example-authentication_{context}[Preparing TLS connectivity for a CLI Kafka client session]
+. xref:proc-oauth-authorization-keycloak-example-check_{context}[Checking authorized access to Kafka using a CLI Kafka client session]
+
+.Connection to Keycloak
+Kafka and Kafka clients must connect to Keycloak for authorization to work. Connection to Keycloak is indirect.
+The connection routes through an ingress HTTP gateway using a hostname on port 443. Kafka clients use the same hostname and port as the Kafka brokers.
+The gateway is part of the Kubernetes infrastructure. This example assumes that Keycloak is deployed using the Keycloak Operator.
+The Keycloak Operator exposes Keycloak over TLS using the ingress HTTP gateway.
+The ingress hostname must be visible to the pods within the cluster for the example to work.
+If it's not visible, try deploying Keycloak to a separate Kubernetes cluster.
+The pods can then access Keycloak using the external hostname to make the connection.
```

Comment:
Feel free to postpone it to separate PR if it was there already before - from the PR it is not clear if the content just moved or is new. 

But I do not think this makes sense. If I run Kafka cluster and clients only inside the same Kubernetes cluster, why would I need to expose Keycloak externally or deploy it in a separate Kubernetes cluster? I think this should just say something like ...

```
The Kafka brokers need to use the same address to access the Keycloak server as the Kafka clients.
For example, if the Kafka clients use an external address based on an Ingress or Route, the Kafka broker should use the same.
```

However, I'm not sure how this applies to authorization. I think it applies only to authentication. So it probably should not be here at all because it will confuse things: `authorization` is configured only once per cluster and needs to use the same address for all clients. `authentication` is configured per-listener and might need to use completely different address for each listener. So maybe we should just remove it from here?

Reply:
@scholzj Thanks. It's been rewritten since the last publication. It's to explain how Keycloak Operator exposes Keycloak before we get into the procedures, as we're assuming Keycloak is deployed that way. 
@mstruk Hi Marko. Jakub suggests we reduce this Keycloak connection section.

Reply:
@scholzj It applies to authorizazion as well since the 'keycloak' authorization has to use the same token endpoint to retrieve a list of grants for the token, and access it in the same way it was accessed when obtaining the token.

But maybe the text could be slightly specific to authorization:

Keycloak authorization uses the token that passed authentication validation checks to obtain from Keycloak a list of grants for the user. Thus, it needs to connect to Keycloak, and it needs to use the same hostname and port that were used when obtaining the token for the authentication. See the `Connection to Keycloak` chapter for 'oauth' authentication about Keycloak Ingres hostname visibility to pods in the cluster.